### ISSUE TITLE
ENH: Flatpak package

### ISSUE THREAD

**Issue by jakobjakobson13** [frescobaldi/frescobaldi#1182]
Dear developers,

could you please provide a flatpak package of frescobaldi on [Flathub](https://flathub.org)?

Regards

**Comment by uliska** [comment]
So far I haven't used Flat packages (except when necessary) because they are so (unnecessarily?) big. And I don't have any idea what it takes to provide an application as a flatpak package.

However, I think it would be good to discuss the issue before the next release. If it is a way to bring Frescobaldi to a wider audience then why not ...

@jakobjakobson13 you could help us by providing information (even if you have to read into the matter yourself) on what it takes.

**Comment by jakobjakobson13** [comment]
Well, so far I got here, ut there are some issues (ssee the comments)
```yaml
---
app-id: org.frescobaldi.frescobaldi.org # can be altered
runtime: org.kde.Platform # perhaps freedesktop.platform would suffice
runtime-version: '5.12'
sdk: org.kde.Sdk
command: frescobaldi
finish-args:
  - --socket=x11
  - --share=ipc
  - --device=dri
  - --socket=wayland
modules:
- name: frescobaldi # so far fresscobaldi can't import sip! changing the order of the modules could help?
  buildsystem: simple
  build-commands:
  - python3 setup.py install --prefix=${FLATPAK_DEST} --root=/ # not 100% sure about the prefix and root option
  sources:
  - type: archive
    url: https://github.com/frescobaldi/frescobaldi/releases/download/v3.0.0/frescobaldi-3.0.0.tar.gz
    sha256: 5fa8e4e356cd7f0fb7b8e83870914423e7fec45fa6e0f8243d7946b9e4829895
- name: openjpeg # copy/pasted
  buildsystem: cmake-ninja
  config-opts:
    - -DBUILD_CODEC=OFF
  cleanup:
  - /lib/openjpeg-2.3
  sources:
  - type: archive
    url: https://github.com/uclouvain/openjpeg/archive/v2.3.0.tar.gz
    sha256: 3dc787c1bb6023ba846c2a0d9b1f6e179f1cd255172bde9eb75b01f1e6c7d71a
- name: poppler # copy/pasted
  buildsystem: cmake-ninja
  config-opts:
    - -DCMAKE_INSTALL_LIBDIR=/app/lib
    - -DCMAKE_INSTALL_INCLUDEDIR=/app/include
  cleanup:
    - /bin
  sources:
  - type: archive
    url: https://poppler.freedesktop.org/poppler-0.80.0.tar.xz
    sha256: 4d3ca6b79bc13b8e24092e34f83ef5f387f3bb0bbd7359a6c078e09c696d104f
- name: pyqt5 # copy/pasted
  config-opts:
    - --disable-static
    - --enable-x11
  sources:
  - type: archive
    url: https://www.riverbankcomputing.com/static/Downloads/PyQt5/5.13.1/PyQt5_gpl-5.13.1.tar.gz
    sha256: 54b7f456341b89eeb3930e786837762ea67f235e886512496c4152ebe106d4af
  - type: script
    commands:
    - python3 configure.py --assume-shared --confirm-license --no-designer-plugin
      --no-qml-plugin --sysroot=/app --qsci-api --qsci-api-destdir=/app/qsci --sipdir=/app/share/sip
      --sip=/app/bin/sip --sip-incdir=/app/include QMAKE_CFLAGS_RELEASE='-I/usr/include/python3.7m/'  QMAKE_CXXFLAGS_RELEASE='-I/usr/include/python3.7m/'
    dest-filename: configure
  modules:
  - name: sip # copy/pasted
    sources:
    - type: archive
      url: https://www.riverbankcomputing.com/static/Downloads/sip/4.19.19/sip-4.19.19.tar.gz
      sha256: 5436b61a78f48c7e8078e93a6b59453ad33780f80c644e5f3af39f94be1ede44
    - type: script
      commands:
      - python3 configure.py --sip-module PyQt5.sip -b ${FLATPAK_DEST}/bin -d ${FLATPAK_DEST}/lib/python3.7/site-packages
        -e ${FLATPAK_DEST}/include -v ${FLATPAK_DEST}/share/sip --stubsdir=${FLATPAK_DEST}/lib/python3.7/site-packages
      dest-filename: configure
    cleanup:
      - /bin
      - /include
- name: python-ly
  buildsystem: simple
  build-commands:
  - python3 setup.py install --prefix=--prefix=${FLATPAK_DEST} --root=/ # not sure about the prefix and root option
  sources:
  - type: archive
    url: https://github.com/frescobaldi/python-ly/archive/v0.9.5.tar.gz
    sha256: 11d2eb5c0f8726d4d8b48bfd104ba30bee1d3c64915028e6710e6e239d1f4c1f
- name: python-poppler-qt5
  buildsystem: simple
  build-commands:
  - python3 setup.py install --prefix=--prefix=${FLATPAK_DEST} --root=${FLATPAK_DEST}/share/sip # not sure about the prefix and root option
  sources:
  - type: archive
    url: https://github.com/frescobaldi/python-poppler-qt5/archive/v0.74.0.tar.gz
    sha256: 8e073ec496a1c39btbe14b53d70c0a8628b7ec3fe0f8db0709440a6a6b553109f
- name: lilypond
  buildsystem: simple
  build-commands:
  - install -D lilypond-2.18.2-1.linux-64.sh /app/bin/lilypond
  sources:
  - type: file
    url: http://lilypond.org/download/binaries/linux-64/lilypond-2.18.2-1.linux-64.sh
    sha256: 4132d13f700f18bd5759837ebe1a44c2bb24205ddf145c920df050b06f8b3d22
```

In the end the start of frescobaldi falis to a missing sip module. Either the build order plays an important role or the pythno path has to be further adapted. I'll try further testing at the week end.

**Comment by fedelibre** [comment]
@jakobjakobson13 Well done!

There's a typo in the SHA256 of python-poppler-qt5. Correct string is `8e073ec496a1c39bbe14b53d70c0a8628b7ec3fe0f8db0709440a6a6b553109f`

I've run flatpak-builder successfully, but I'm getting the same error about missing sip.
I've seen this before. I think that running `make && make install` right after `python3 configure.py` (for the sip module) might fix the issue. Not sure how to add this in the Flatpak manifest.

BTW, before the summer I tried to build wheels of python-poppler-qt5 but I found a problem and then had no time to go on. I put some notes on the wiki:

- https://github.com/frescobaldi/python-poppler-qt5/wiki/Build-wheels-in-CentOS-7
- https://github.com/frescobaldi/python-poppler-qt5/wiki/Build-wheels-using-manylinux

**Comment by jakobjakobson13** [comment]
> There's a typo in the SHA256 of python-poppler-qt5. Correct string is `8e073ec496a1c39bbe14b53d70c0a8628b7ec3fe0f8db0709440a6a6b553109f`
> 

You are right.


> Not sure how to add this in the Flatpak manifest.
>

You just add the lines
```
- make
- make install
```
right after `python3 configure.py` (for the sip module) with the same intendation level as the `python3 configure.py` line.

**Comment by fedelibre** [comment]
Adding those lines didn't work.

However, it seems that sip is no more required as PyQt5 uses its own private sip. And Wilbert has adapted Frescobaldi to this, see 92abb8dc7d0522166fdd90fe9921194678e229b1

So I'm now trying to build a flatpak with Frescobaldi from git master.

**Comment by fedelibre** [comment]
@jakobjakobson13 I've almost succeeded. Frescobaldi runs successfully, but LilyPond doesn't work.
If I keep the install command you used, lilypond doesn't find any argument: even `lilypond --version` doesn't work. So I've tried `sh lilypond.sh --prefix ...` but there's something wrong with permissions.
Note also that we need all the binaries produced by a lilypond installation (convert-ly, etc.).

Find attached my current manifest.
I'm going to work now, feel free to test it further. Thanks!
[org.frescobaldi.Frescobaldi.yml.txt](https://github.com/frescobaldi/frescobaldi/files/3668970/org.frescobaldi.Frescobaldi.yml.txt)

BTW, I think a flatpak from master is more interesting than the old stable.
One of the cool things about Flatpak is that you can have concurrent installations of several branches of the same app.

**Comment by fedelibre** [comment]
I found a workaround, using chmod to change the permission of a file.
Now I have a working lilypond executable and all the other lilypond commands.

```
- name: lilypond
  buildsystem: simple
  build-commands:
  - sh lilypond-2.18.2-1.linux-64.sh --prefix ${FLATPAK_DEST}
  - chmod 755 ${FLATPAK_DEST}/lilypond/usr/lib/libpython2.4.so.1.0
  sources:
  - type: file
    url: http://lilypond.org/download/binaries/linux-64/lilypond-2.18.2-1.linux-64.sh
    sha256: 4132d13f700f18bd5759837ebe1a44c2bb24205ddf145c920df050b06f8b3d22
```
It will need some testing.
For example, the variables in the Snippets (like `$(LILYPOND_VERSION)`) do not work.
And we'll probably have to enable the access to network sooner or later, as Frescobaldi is going to need it soon, I think.

**Comment by jakobjakobson13** [comment]
> I found a workaround, using chmod to change the permission of a file.
> Now I have a working lilypond executable and all the other lilypond commands.
> 
> ```
> - name: lilypond
>   buildsystem: simple
>   build-commands:
>   - sh lilypond-2.18.2-1.linux-64.sh --prefix ${FLATPAK_DEST}
>   - chmod 755 ${FLATPAK_DEST}/lilypond/usr/lib/libpython2.4.so.1.0
>   sources:
>   - type: file
>     url: http://lilypond.org/download/binaries/linux-64/lilypond-2.18.2-1.linux-64.sh
>     sha256: 4132d13f700f18bd5759837ebe1a44c2bb24205ddf145c920df050b06f8b3d22
> ```
> 
> It will need some testing.

Great!

> For example, the variables in the Snippets (like `$(LILYPOND_VERSION)`) do not work.

Is there anything specific I can do?

> And we'll probably have to enable the access to network sooner or later, as Frescobaldi is going to need it soon, I think.

This can easily be done via the [finish-args](http://docs.flatpak.org/en/latest/sandbox-permissions.html).

**Comment by fedelibre** [comment]
> > For example, the variables in the Snippets (like `$(LILYPOND_VERSION)`) do not work.
> 
> Is there anything specific I can do?

Find why it's not working?

I'm not a developer. I can only guess that this is handled by python. Perhaps the python distributed in the runtime is missing some module which handles this sort of stuff? (though it looks like a basic function)
@uliska Perhaps you can give some hint about the code which expands the variables in the snippets tool?

Also, it would be better to start working on this manifest file in a branch and create a PR, as I think this should be managed upstream (and then used by any flatpak distributor out there). I'd be happy to maintain it. Maybe @jakobjakobson13 could help also?
Before creating the branch, we should discuss where this file should be placed. This is a question I made to myself for other desktop thinkgs (like AppData and .desktop files). I'll create a new issue to discuss this and link to all the related issues.

**Comment by fedelibre** [comment]
Well,  it may be something related to how Frescobaldi determines the version of currently installed lilypond. In fact, it seems it's not aware of the version, as you can see on these screenshots:

![fresco_missing_lily_version](https://user-images.githubusercontent.com/1644647/66081538-798aa680-e568-11e9-909c-ee9b2aba2f84.png)
![fresco_missing_lily_version_2](https://user-images.githubusercontent.com/1644647/66081541-798aa680-e568-11e9-8a12-ad6b6f632df0.png)

**Comment by fedelibre** [comment]
Another problem: opening the documentation browser is not possible when using Qt 5.12, because QtWebKit has been deprecated. I might try Qt 5.8. See [issue 919](https://github.com/frescobaldi/frescobaldi/issues/919).

**Comment by fedelibre** [comment]
Nope. The oldest org.kde.Sdk on Flathub is 5.9. I guess we'll have to wait for a fix to issue #919

**Comment by uliska** [comment]
>> @uliska Perhaps you can give some hint about the code which expands the variables in the snippets tool?

>Well, it may be something related to how Frescobaldi determines the version of currently installed lilypond. 

Definitely. Code in `snippet/expand.py` shows that `LILYPOND_VERSION` is expanded to the result of calling `lilypondinfo.preferred().versionString()`. So if your installation does not have this available it is clear that it does not work in snippets.

However, I don't have any idea how this relates to the issue of using Fresocbaldi from a flatpak.

**Comment by fedelibre** [comment]
Everything is working fine now. I _think_ it was due to sandbox permissions. I've added `--filesystem=home` and  cleaned the directory `~/.var/app/org.frescobaldi.Frescobaldi`.

**Comment by fedelibre** [comment]
For the records, I've been working on adding a fluidsynth service for the flatpak, in order to enable MIDI playback. As soon as I've solved a problem, I'll try to submit the manifest to Flathub (which might be also a chance to get a proper review from flatpak experts).

**Comment by jakobjakobson13** [comment]
@fedelibre Have you already checked the following manifests for the midi support?
- [Kodi](https://github.com/flathub/tv.kodi.Kodi/blob/master/deps/fluidsynth.json)
- [LMMS](https://github.com/flathub/io.lmms.LMMS/blob/master/io.lmms.LMMS.json)
- [Muse3](https://github.com/flathub/org.musesequencer.Muse3/blob/master/org.musesequencer.Muse3.json)

**Comment by fedelibre** [comment]
Thanks @jakobjakobson13 

I was trying to use flathub shared modules, as they have fluidsynth and portmidi in a single manifest, but I found [this problem](https://github.com/flathub/shared-modules/issues/76).
Muse3 fluidsynth module is working fine.
I'll proceed this way...

Do you know how to add a systemd service in the manifest? Something like [this one](https://github.com/frescobaldi/frescobaldi/wiki/MIDI-playback-on-Linux#systemd-fluidsynth-service).

**Comment by fedelibre** [comment]
BTW, I'm pushing changes to the manifest in [this temporary repository](https://github.com/fedelibre/org.frescobaldi.Frescobaldi). I will delete it once the manifest is ready to be submitted to Flathub.

**Comment by fedelibre** [comment]
> However, it seems that sip is no more required as PyQt5 uses its own private sip. And Wilbert has adapted Frescobaldi to this, see [92abb8d](https://github.com/frescobaldi/frescobaldi/commit/92abb8dc7d0522166fdd90fe9921194678e229b1)

@jakobjakobson13 I've now backported this patch and applied it to the 3.0.0 tarball. Frescobaldi 3.0.0 flatpak is now working correctly. I'll send soon a PR to flathub.

Before doing so, I'll open an issue here to discuss which new screenshots we should use in the appdata file.

**Comment by jakobjakobson13** [comment]
Great. I'm looking forward to it.

**Comment by fedelibre** [comment]
I should never say "soon" these days :-)

Today I made a good find: there's no need to bundle lilypond as we can easily use versions installed on the host (as access to host is allowed in the manifest).
To be more precise, any lilypond installed within the home directory can be installed normally using the regular path, e.g. `~/.local/bin/lilypond`. If you want to use an installation in `/usr`, I'm afraid it's more complicate. You should use a different path: `/var/run/host/usr/bin/lilypond`. But then guile shared libraries cannot be found:

```
/var/run/host/usr/bin/lilypond: error while loading shared libraries: libguile.so.17: cannot open shared object file: No such file or directory
```

**Comment by fedelibre** [comment]
It seems the sandbox can use applications installed in /home and /opt but not in /usr. This means we cannot use LilyPond installations provided by the distributions, but we can install as many versions of the lilypond.org installers.
I'll remove the LilyPond bundle from the manifest.

Hopefully next week I'll send the PR to flathub.

**Comment by jakobjakobson13** [comment]
> I should never say "soon" these days :-)
> 
> Today I made a good find: there's no need to bundle lilypond as we can easily use versions installed on the host (as access to host is allowed in the manifest).
> To be more precise, any lilypond installed within the home directory can be configured normally in Frescobaldi using the regular path, e.g. `~/.local/bin/lilypond`. If you want to use an installation in `/usr`, I'm afraid it's more complicate. You should put a different path in Frescobaldi configuration: `/var/run/host/usr/bin/lilypond`. But then guile shared libraries cannot be found:
> 
> ```
> /var/run/host/usr/bin/lilypond: error while loading shared libraries: libguile.so.17: cannot open shared object file: No such file or directory
> ```

Are you sure that you want to exclude the LIlyPond files from the flatpak package? Flatpak packages are supposed to be out-of-the-box solutions without any external dependencies and further installing/downloading any further files.

**Comment by uliska** [comment]
Am 24. November 2019 12:19:36 MEZ schrieb jakobjakobson13 <notifications@github.com>:
>> I should never say "soon" these days :-)
>> 
>> Today I made a good find: there's no need to bundle lilypond as we
>can easily use versions installed on the host (as access to host is
>allowed in the manifest).
>> To be more precise, any lilypond installed within the home directory
>can be configured normally in Frescobaldi using the regular path, e.g.
>`~/.local/bin/lilypond`. If you want to use an installation in `/usr`,
>I'm afraid it's more complicate. You should put a different path in
>Frescobaldi configuration: `/var/run/host/usr/bin/lilypond`. But then
>guile shared libraries cannot be found:
>> 
>> ```
>> /var/run/host/usr/bin/lilypond: error while loading shared libraries:
>libguile.so.17: cannot open shared object file: No such file or
>directory
>> ```
>
>Are you sure that you want to exclude the LIlyPond files from the
>flatpak package? Flatpak packages are supposed to be out-of-the-box
>solutions without any external dependencies and further
>installing/downloading any further files.

Yes, depending on the user having installed LilyPond, and in /home, goes against the Flatpak idea.

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by fedelibre** [comment]
You are right. Actually I want to bundle the stable version. But being able to run also a development version installed in /home is convenient, especially considering the slow release cycle of LilyPond (stable).

**Comment by fedelibre** [comment]
For the records, I managed to build lilypond and now I have a flatpak mostly working.
There's only one issue left: the documentation browser is not working because I haven't built qtwebengine yet.

I'm going in holidays but I will work on it in September.
So hopefully we'll have a Frescobaldi flatpak quite soon.

**Comment by fedelibre** [comment]
I think we are close to the publication on Flathub. The latest build was successful. You can test it:

    flatpak install --user https://dl.flathub.org/build-repo/28366/org.frescobaldi.Frescobaldi.flatpakref

There's only one missing feature: documentation browser is not working. Problem described [here](https://github.com/flathub/flathub/pull/1277#issuecomment-691146556).

lilypond 2.20 is included.

**Comment by fedelibre** [comment]
Frescobaldi is now available on Flathub!
https://flathub.org/apps/details/org.frescobaldi.Frescobaldi

I'm closing this issue.

In the next days I'll send a PR  to revise the metainfo file and add the news in the Changelog.
